File: Yandex Algorithm 2013/Test Round/F/main.py
# reading is too slow
IN = [line for line in open("taxi.in")]

M, D, N = [int(x) for x in IN[0].split()]

X = [int(x) for x in IN[1].split()]

Y = []
need = -1

for v in X:
    if v >= M - D and (v < need or need == -1):
        need = v

enough = need

for v in X:
    if v == need:
        need = -1
        continue
    Y += [v]

X = Y
X.sort()

position = 0

OUT = open('taxi.out', 'w')

answer = 0
while position < M:
    if len(X) == 0 and enough == -1:
        print("0", file=OUT)
        exit()
    v = None
    need = M - position
    if position < D:
        need += (D - position)
    
    if len(X) > 0 and enough < need:
        v = X.pop()
    else:
        v = enough
        enough = -1

    if position < D:
        v -= D - position
    else:
        v -= position - D
    if v <= 0:
        continue
    position += v
    answer += 1

print(answer, file=OUT)

File: Yandex Algorithm 2013/Test Round/F/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    ifstream cin("taxi.in");
    ofstream cout("taxi.out");

    int64_t M, D; int N;
    cin >> M >> D >> N;

    vector<int64_t> X(N);
    for (int i = 0; i < N; ++i)
        cin >> X[i];

    vector<int64_t> Y;
    int64_t need = -1;

    for (auto &v : X)
        if (v >= M - D and (v < need or need == -1))
            need = v;

    int64_t enough = need;

    for (auto &v : X)
        if (v == need) {
            need = -1;
            continue;
        } else
            Y.push_back(v);
    swap(X, Y);
    sort(X.begin(), X.end());

    int64_t position = 0;

    int answer = 0;
    while (position < M) {
        if (X.size() == 0 and enough == -1) {
            cout << "0\n";
            return 0;
        }

        int64_t v = 0;
        need = M - position;
        if (position < D)
            need += (D - position);
        if (X.size() and enough < need) {
            v = X.back();
            X.pop_back();
        } else {
            v = enough;
            enough = -1;
        }

        if (position < D)
            v -= D - position;
        else
            v -= position - D;
        if (v <= 0)
            continue;
        position += v;
        answer++;
    }

    cout << answer << "\n";
}
